File: Chapter 09/Python/section_9_3_4_serverless_functions/infrastructure/web/mock_http_request.py
from rx import of
from ...infrastructure.web.mock_webhook_payload import MockWebhookPayload

class MockHttpRequest:
    """
    THE WEB STANDARDS CONTRACT (Standard Reactive HTTP Request):
    
    DESIGN NOTE:
    This class simulates a standard HTTP Request object (like those found in 
    Flask, FastAPI, or Django). It contains the 'MockWebhookPayload'.
    
    ARCHITECTURAL CRITIQUE:
    This is the "Clarity Peak." Because Webhooks interact with your code 
    using the "Language of the Web" (HTTP/JSON) rather than "Vendor SDKs," 
    this infrastructure mock is the most reusable. By wrapping the 
    payload in a standard Request object, we demonstrate that our handler 
    isn't a "Cloud Function"—it's a standard web endpoint that happens 
    to be triggered by a database. This represents the ultimate decoupling 
    on the Compute Spectrum.
    """
    def __init__(self, payload: MockWebhookPayload):
        self._payload = payload
        self.method = "POST"
        self.headers = {
            "Content-Type": "application/json",
            "User-Agent": "Supabase-Webhook-Sim"
        }

    def get_body_stream(self):
        """
        Returns the request body as a Reactive Observable stream.
        
        TEACHING NOTE:
        Modeling the body as an Observable allows our logic to remain 
        asynchronous and projective, consistent with the AWS and Azure 
        implementations, but using standard pipes.
        """
        return self._payload.get_body_stream()

File: Chapter 09/C#/section_9_3_4_serverless_functions/Infrastructure/Web/MockHttpRequest.cs
using System;
using System.Text.Json;
using System.Reactive.Linq;

namespace Chapter09.ServerlessFunctions.Infrastructure.Web;

/// <summary>
/// THE INFRASTRUCTURE WRAPPER (Standard HTTP Request):
/// 
/// TEACHING NOTE:
/// This class simulates a standard HTTP Request object (the kind you would find 
/// in ASP.NET Core). By wrapping the payload here, we demonstrate that 
/// Webhooks interact with your application logic using the "Language of the Web" 
/// (HTTP/JSON) rather than "Vendor-Specific Dialects." This makes the 
/// entry point into your system much more predictable and easier to test.
/// </summary>
public class MockHttpRequest
{
    private readonly string _jsonBody;

    public MockHttpRequest(MockWebhookPayload payload)
    {
        _jsonBody = JsonSerializer.Serialize(payload);
    }

    /// <summary>
    /// Returns the request body as a Reactive Observable Stream.
    /// 
    /// ARCHITECTURAL NOTE:
    /// Modeling the request body as a stream allows our logic to handle 
    /// data reactively. Even if the body is a single object, we treat it 
    /// as a stream to maintain a consistent 'Reactive Pipeline' across 
    /// all our cloud implementations.
    /// </summary>
    /// <returns>An IObservable emitting the deserialized payload.</returns>
    public IObservable<MockWebhookPayload> GetBody()
    {
        var payload = JsonSerializer.Deserialize<MockWebhookPayload>(_jsonBody, 
            new JsonSerializerOptions { PropertyNameCaseInsensitive = true })!;
        return Observable.Return(payload);
    }
}